### DISCUSSION TITLE
Simultaneous Events Execution

### DISCUSSION THREAD

**Discussion by dvilchez** [cypress-io/cypress#30702] [Questions and Help] [answered]
Hi folks,

I'm curious to know if there's an existing method to send multiple gestures simultaneously using the trigger API, rather than sending them one by one. We're aiming to replay complex gestures, such as pinch to zoom or a long two-finger scroll, in a timely manner. Our attempts with the trigger API have been quite slow for these gestures.

If anyone has any insights or alternative suggestions, I would greatly appreciate your input. Thank you!

**Comment by portyu9** [comment, marked as answer]
Hi @dvilchez 

There is no Cypress `trigger()` API that executes two independent Cypress commands **simultaneously**.

Cypress commands are intentionally queued and executed serially.

For a gesture such as pinch-to-zoom, however, the important distinction is whether you are testing:

1. **your application's gesture event handler**, or
2. the browser's **real native pinch/scroll behavior**

For an application-level handler, a synthetic `TouchEvent` can contain **multiple active touch points in one event**. That is closer to how a pinch gesture is represented than trying to run two `.trigger()` commands at exactly the same time.

For example, in a browser that exposes the `Touch`/`TouchEvent` constructors, you can dispatch a single event containing two touches and then subsequent `touchmove` events with updated coordinates.

However, synthetic touch events have an important limitation: they are **untrusted events**. Browsers do not perform protected/native default actions such as real browser pinch-zoom simply because JavaScript dispatched a synthetic `TouchEvent`.

The same limitation applies to using `.trigger()`:

```js
cy.get(...).trigger('touchmove')
```

It can invoke your application's event listeners, but it does not reproduce all native browser behavior.

So:

- **testing your own gesture code** → synthesize touch/pointer events representing multiple contacts
- **testing real browser pinch zoom / native two-finger scrolling** → use native browser/device automation rather than `cy.trigger()`

Also note that `TouchEvent` construction itself has browser compatibility limitations, so if your application supports Pointer Events, synthetic `PointerEvent`s with distinct `pointerId`s may be easier for application-level gesture tests.

Please mark this answer as accepted if it helped, thank you.
